### ISSUE TITLE
Poor TCP Load Balancing

### ISSUE THREAD

**Issue by dadgar** [traefik/traefik#7863]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I have three instances of a service that are registered in a Consul Catalog as TCP services. They all have the following tags:

```yml
          "traefik.enable=true",
          "traefik.tcp.routers.scada.entrypoints=scada",
          "traefik.tcp.routers.scada.rule=HostSNI(`*`)",
          "traefik.tcp.routers.scada.tls=true",
          "traefik.tcp.routers.scada.tls.passthrough=true",
```

Relevant part of Traefik config:

```toml
# Load services from Consul catalog
[providers]
  [providers.consulCatalog]
    # Do not automatically expose our services.
    exposedByDefault = false

    # Poll Consul every 5 seconds
    refreshInterval = "5s"

    # Address to talk to Consul
    [providers.consulCatalog.endpoint]
      address = "XXX:YYY"

[entryPoints]
  [entryPoints.scada]
    address = "0.0.0.0:7223"
```

 I have client connecting at a very steady rate to the endpoint.

### What did you expect to see?

I would expect to see a very even balance of connections to the three services.

### What did you see instead?

So when I do a rollout of the service they start out with an even load but over time the load distribution becomes heavily weighted to the first service instance return from the Consul Catalog, followed by the second, then third. You can see that in this graph here which tracks the number of open connections by instance:

<img width="2216" alt="Screen Shot 2021-02-05 at 6 05 57 PM" src="https://user-images.githubusercontent.com/1047914/107105824-24eedf80-67dd-11eb-8b34-3b2ac71045c0.png">

The next graph shows new connections being established over time group by instance type. You can see it is essentially all going to a single instance:

<img width="2214" alt="Screen Shot 2021-02-05 at 6 19 55 PM" src="https://user-images.githubusercontent.com/1047914/107106116-cfb3cd80-67de-11eb-8ccb-a1c234c4ef06.png">

Thus as clients disconnect naturally, all new connections keep bunching up on just one instance.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
v2.1.8
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Running on Nomad on AWS.

### Hypothesis

<img width="834" alt="Screen Shot 2021-02-05 at 6 06 17 PM" src="https://user-images.githubusercontent.com/1047914/107105854-6a131180-67dd-11eb-9982-a3e5bc2939a2.png">

Digging through the code a bit, my hypothesis is that every five seconds when the Consul catalog is pulled and passed as a new config, a new router is created here:

https://github.com/traefik/traefik/blob/d211437d6ca4b26bd5e8f7901adb916bdf44f61b/cmd/traefik/traefik.go#L284

Following the following code bits:

https://github.com/traefik/traefik/blob/d211437d6ca4b26bd5e8f7901adb916bdf44f61b/pkg/server/routerfactory.go#L80-L81
https://github.com/traefik/traefik/blob/d211437d6ca4b26bd5e8f7901adb916bdf44f61b/pkg/server/router/tcp/router.go#L80
https://github.com/traefik/traefik/blob/d211437d6ca4b26bd5e8f7901adb916bdf44f61b/pkg/server/router/tcp/router.go#L242
https://github.com/traefik/traefik/blob/d211437d6ca4b26bd5e8f7901adb916bdf44f61b/pkg/server/service/tcp/service.go#L68

It appears the end result is the weighted round robin load balancer gets reset every 5 seconds to the same order of services returned by the Consul Catalog.

**Comment by ldez** [comment]
Hello,

have you tried with a recent version (v2.4.2)?

**Comment by dadgar** [comment]
Hey @ldez, I have not tried on the latest version. Has anything changed to expect an improvement?

For what it is worth, the code paths I linked are on the latest commit.

**Comment by dadgar** [comment]
@ldez @jbdoumenjou Let me know if you need any more information or re-pro steps.

**Comment by ldez** [comment]
The WRR is reset at every reload of the dynamic configuration, but the refresh occurs only if the new configuration is different from the old one.

How many changes do you have?
How many Traefik are you using?

**Comment by dadgar** [comment]
Hey I have shared a screenshot of Traefik's metric for reloading. Is that the best metric to share? There are 8 instances of Traefik all watching the same Catalog. It looks like each instance is refreshing 12 times a minute which lines up with the 5 second refetch configuration.

<img width="1267" alt="Screen Shot 2021-02-10 at 9 56 41 AM" src="https://user-images.githubusercontent.com/1047914/107550831-5d4c4000-6b86-11eb-9184-a536c315d3bc.png">


The catalog is only updated when we deploy a service which on average probably once every hour during business hours.

**Comment by mmatur** [comment]
Hi @dadgar ,

Thanks for the information. 

Following all the information that you gave us, we are not able to identify why the configuration is refresh every 5s without any change on the catalog. 

Are you able to start traefik in [debug mode](https://doc.traefik.io/traefik/observability/logs/#level) and share with us logs?

**Comment by dadgar** [comment]
@mmatur Sure, I have the logs. Is there a way to share privately?

**Comment by mmatur** [comment]
Hi @dadgar,

Could you please send me a DM on our [community forum](https://community.traefik.io/u/michael/summary)?

**Comment by dadgar** [comment]
@mmatur Signed up for an account but can't see a way to DM you. Feel free to DM me and I can reply https://community.traefik.io/u/dadgar or send me an email at alex@hashicorp.com

**Comment by ldez** [comment]
@dadgar I think I have found the root cause, could you try with a more recent version at least the v2.2.2 but the v2.4.2 is better.

**Comment by dadgar** [comment]
Hey @ldez @mmatur,

Tried out v2.4.2 and it is much better! Thank you so much. We had to roll it back due to some errors we were getting in some other middleware config but while we had it running the metrics looked much better:

Traffic was more evenly spread, can see the before and after:
<img width="827" alt="Screen Shot 2021-02-12 at 9 14 14 AM" src="https://user-images.githubusercontent.com/1047914/107812326-5d824200-6d24-11eb-9daa-3a626cd1600f.png">

Traefik stopped reloading constantly!
<img width="736" alt="Screen Shot 2021-02-12 at 9 14 54 AM" src="https://user-images.githubusercontent.com/1047914/107812334-5eb36f00-6d24-11eb-8c9e-cc3376fd9a13.png">

I also saw many log lines saying `Skipping same configuration` for the consul catalog! Going to close and I can re-open if we do see the same problem once we fully switch to 2.4.2
